### PR TITLE
SNOW-975597: use result scan for table sproc result

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-975597

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/stored_procedure.py [hunk 1]
```diff
@@ -102,8 +103,13 @@ def __call__(
 
         if self._anonymous_sp_sql:
             call_sql = generate_call_python_sp_sql(session, self.name, *args)
-            df = session.sql(f"{self._anonymous_sp_sql}{call_sql}")
+            query = f"{self._anonymous_sp_sql}{call_sql}"
+            df = session.sql(query)
             if self._is_return_table:
+                qid = session._conn.execute_and_get_sfqid(
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/stored_procedure.py:109]: Does the same query get run twice?

File: src/snowflake/snowpark/_internal/server_connection.py [hunk 2]
```diff
@@ -339,6 +339,22 @@ def notify_query_listeners(self, query_record: QueryRecord) -> None:
...
+        statement_params: Optional[Dict[str, str]] = None,
+    ) -> str:
+        if statement_params is None:
+            statement_params = {}
+        if is_ddl_on_temp_object:
+            statement_params["SNOWPARK_SKIP_TXN_COMMIT_IN_DDL"] = True
+        results_cursor = self._cursor.execute(query, _statement_params=statement_params)
+        self.notify_query_listeners(
+            QueryRecord(results_cursor.sfqid, results_cursor.query)
+        )
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/server_connection.py:355]: Curious: is there a good way to reduce the duplication if `statement_params` and query listener's code? I would guess this code block is quite reusable.

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/stored_procedure.py [hunk 3]
```diff
@@ -102,8 +103,13 @@ def __call__(
...
+                    query, statement_params=statement_params
+                )
+                df = session.sql(result_scan_statement(qid))
```
Comment (sfc-gh-mkeller) [inline src/snowflake/snowpark/stored_procedure.py:112]: The return-type of this function is incorrect. It should be `Any` and not `any`. The lower-case version is a builtin. I know this isn't your doing, but could you please fix it?

File: src/snowflake/snowpark/stored_procedure.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/stored_procedure.py:109]: `df = session.sql(query)` does not run the query, doing an internal collect or running function `execute_and_get_sfqid` does. I see df on line 107 is only used for the final return on line 114. I'll update the PR and move it so it become less confusing.

Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/stored_procedure.py:109]: I understand now. Thanks for explaining.

File: src/snowflake/snowpark/_internal/server_connection.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/server_connection.py:355]: I think I can remove extra processing done on statement params are cursor is happy with accepting a None statement param. I'll remove ddl on temp object to keep things simple as it does not have any usage now. seeing other occurrences of statement params, does not seem like there is much duplication.

I'll create a helper for execute and notify query listener as I see them being used together frequently. Do you want me to refactor them in this PR or should I create a follow-up JIRA for that

File: tests/integ/test_udf.py [hunk 4]
```diff
@@ -1781,7 +1781,12 @@ def return_type_in_dataframe(x):
         (
             IntegerType,
             [[1048576]],
-            ("<class 'numpy.int32'>", "<class 'numpy.int64'>", "<class 'int'>", "<class 'numpy.intc'>"),
+            (
+                "<class 'numpy.int32'>",
+                "<class 'numpy.int64'>",
+                "<class 'int'>",
+                "<class 'numpy.intc'>",
+            ),
```
Comment (sfc-gh-aalam) [inline tests/integ/test_udf.py:1789]: lint fix

**Pull request merged**